podcasts/Sharing/Clip/VideoExporter.swift

```diff
@@ -0,0 +1,203 @@
+        progress.totalUnitCount = Int64(loopFrameCount) + parameters.additionalLoadingCount
+
+        try await withTaskCancellationHandler {
+            // Export initial seconds long video
+            try await exportInitialVideo(of: view, to: temporaryFileURL, with: parameters, frameCount: loopFrameCount, progress: progress)
+            FileLog.shared.addMessage("VideoExporter Initial Video Ended: \(start.timeIntervalSinceNow)")
+
+            // Create & export final composition at full length by concatenating seconds long video from above until duration
+            let composition = try await createFinalComposition(from: temporaryFileURL, with: parameters, to: outputURL, progress: progress)
+            try await exportFinalComposition(composition, to: outputURL, duration: parameters.duration, fileType: parameters.fileType, progress: progress)
+
+            // Clean up temporary file
+            try? FileManager.default.removeItem(at: temporaryFileURL)
+            progress.completedUnitCount = progress.totalUnitCount
+            FileLog.shared.addMessage("VideoExporter Ended: \(start.timeIntervalSinceNow)")
+        } onCancel: {
+            progress.cancel()
+        }
+    }
+
+    //MARK: - Private Methods
+
+    // Step 1 of video export
+    private static func exportInitialVideo<Content: AnimatableContent>(of view: Content, to outputURL: URL, with parameters: Parameters, frameCount: Int, progress: Progress) async throws {
+        let videoWriter = try AVAssetWriter(outputURL: outputURL, fileType: parameters.fileType)
+        let videoSettings: [String: Any] = [
+            AVVideoCodecKey: AVVideoCodecType.h264,
+            AVVideoWidthKey: parameters.size.width,
+            AVVideoHeightKey: parameters.size.height
+        ]
+        let videoWriterInput = AVAssetWriterInput(mediaType: .video, outputSettings: videoSettings)
+        let adaptor = AVAssetWriterInputPixelBufferAdaptor(assetWriterInput: videoWriterInput, sourcePixelBufferAttributes: nil)
+
+        videoWriter.add(videoWriterInput)
+        videoWriter.startWriting()
+        videoWriter.startSession(atSourceTime: .zero)
+
+        try await writeFrames(of: view,
+                              size: parameters.size,
+                              fps: parameters.fps,
+                              videoWriterInput: videoWriterInput,
+                              videoWriter: videoWriter,
+                              adaptor: adaptor,
+                              progress: progress,
+                              frameCount: frameCount)
+    }
+
+    // Part of Step 1
+    private static func writeFrames<Content: AnimatableContent>(of view: Content, size: CGSize, fps: Int, videoWriterInput: AVAssetWriterInput, videoWriter: AVAssetWriter, adaptor: AVAssetWriterInputPixelBufferAdaptor, progress: Progress, frameCount: Int) async throws {
+        let counter = Counter()
+        try await videoWriterInput.unsafeRequestMediaDataWhenReady { continuation in
```

Comment:
The `requestMediaDataWhenReady` isn't great for async/await since the block can be called repeatedly but the block needs to run synchronously producing frames. At the same time, it needs to end when the frames are done, so needs some context of an external frame count. I've added a "counter" actor to track the frame count and `unsafeWait` to synchronously block on the async `pixelBuffer` call. Seemed to work well in testing and there shouldn't be many of these running simultaneously.